File: web/src/v2/core/llm.ts
import { useEffect } from 'react'
import { create } from 'zustand'

/**
 * What the language models are producing, right now.
 *
 * A second `EventSource`, on `/llm`, deliberately separate from the one in
 * `events.ts`. That stream carries *state*: frames are coalesced per video and
 * merged last-wins, which is right for a task that moved and wrong for text,
 * where merging two frames loses the first one's words. This carries an
 * append-only log, so it gets its own connection.
 *
 * It also gets its own lifetime. The console is the only thing that reads it,
 * so the connection opens when the console is shown and closes when it is
 * hidden — and the server retains recent exchanges, so reopening replays what
 * was missed rather than starting blank. A panel nobody is looking at costs
 * nothing at all, which is the whole reason this is not folded into the stream
 * that must always be connected.
 *
 * Not in the query cache, for the reason the scheduler snapshot is not: there
 * is no endpoint behind it that anything fetches, so a cache entry would be a
 * fetch that never happens wrapped around a push that always does.
 */

/** One frame, as the server sends it. Text is a delta; runs are append-only. */
interface LLMFrame {
  run: number
  videoId: string
  label: string
  model: string
  text?: string
  done?: boolean
  error?: string
  truncated?: boolean
  startedAt: string
  ms?: number
}

/** One exchange, as the console shows it. */
export interface LLMRun {
  run: number
  videoId: string
  label: string
  model: string
  text: string
  done: boolean
  error?: string
  truncated: boolean
  startedAt: string
  ms?: number
}

/**
 * What the client keeps, mirroring what the server retains.
 *
 * Both caps matter and they cap different things. A browser left open across a
 * fifty-chapter render would otherwise accumulate every exchange of it, and a
 * model that has started repeating itself would otherwise grow one of them
 * without limit. Neither is a reason for a log window to become the largest
 * thing in the tab.
 */
const MAX_RUNS = 32
const MAX_RUN_CHARS = 64 * 1024

interface LLMState {
  runs: LLMRun[]
  connected: boolean
}

const useStore = create<LLMState>(() => ({ runs: [], connected: false }))

/** The exchanges to show, oldest first. */
export function useLLMRuns(): LLMRun[] {
  return useStore((s) => s.runs)
}

/** Whether the console is receiving. */
export function useLLMConnected(): boolean {
  return useStore((s) => s.connected)
}

/**
 * Applies one frame.
 *
 * A run it has not seen is a new exchange; one it has appends. That is the same
 * operation for a live frame and for a backlog frame carrying everything so far,
 * which is what lets a console opened halfway through a generation be served by
 * the code that serves one that was open from the start.
 */
function apply(frame: LLMFrame): void {
  useStore.setState((state) => {
    const index = state.runs.findIndex((r) => r.run === frame.run)
    const previous = index === -1 ? undefined : state.runs[index]

    const next: LLMRun = {
      run: frame.run,
      videoId: frame.videoId,
      label: frame.label,
      model: frame.model,
      text: clamp((previous?.text ?? '') + (frame.text ?? '')),
      done: frame.done ?? previous?.done ?? false,
      error: frame.error ?? previous?.error,
      truncated: frame.truncated ?? previous?.truncated ?? false,
      startedAt: frame.startedAt,
      ms: frame.ms ?? previous?.ms,
    }

    if (previous) {
      const runs = state.runs.slice()
      runs[index] = next
      return { runs }
    }
    // Appended, then trimmed from the front: the server sends them in the order
    // they began, and that is the order they are read in.
    return { runs: [...state.runs, next].slice(-MAX_RUNS) }
  })
}

/** Keeps the tail, which is the half being watched. */
function clamp(text: string): string {
  return text.length <= MAX_RUN_CHARS ? text : text.slice(text.length - MAX_RUN_CHARS)
}

/**
 * Connects for as long as the caller is mounted.
 *
 * Mounted by the console and by nothing else. Two consoles would mean two
 * connections writing the same store, which is wasteful rather than wrong — but
 * there is only ever one panel, so it does not arise.
 */
export function useLLMStream(): void {
  useEffect(() => {
    const source = new EventSource('/llm')

    const onFrame = (event: MessageEvent<string>) => {
      try {
        apply(JSON.parse(event.data) as LLMFrame)
      } catch {
        // A malformed frame is dropped, not thrown. The stream outlives any one
        // bad message.
      }
    }

    // Cleared on *every* open, not once on mount. `EventSource` reconnects on
    // its own without this effect running again, and every connection is served
    // the backlog from the beginning — appending that onto runs already applied
    // would print the last few minutes twice. The browser fires this before it
    // delivers a byte of the body, so the reset always lands first.
    source.onopen = () => useStore.setState({ runs: [], connected: true })
    source.onerror = () => useStore.setState({ connected: false })
    source.addEventListener('llm', onFrame as EventListener)

    return () => {
      source.removeEventListener('llm', onFrame as EventListener)
      source.close()
      useStore.setState({ runs: [], connected: false })
    }
  }, [])
}

File: delivery/http/llm.go
package http

import (
	"encoding/json"
	"log/slog"
	"net/http"
	"time"

	"github.com/tbui/yt-studio/domain/provider"
)

// LLMStreamSource is the narrow half of the LLM log this handler needs.
//
// Subscribe hands back the backlog and the live channel in one call rather than
// two, and that is not a convenience. Taken separately, a frame landing between
// them would be absent from the snapshot and never delivered — in append-only
// text that is not a stale reading but a hole in the middle of a sentence.
type LLMStreamSource interface {
	Subscribe() ([]provider.LLMFrame, <-chan provider.LLMFrame, func())
}

// llmFrame is the wire shape of one moment in an exchange.
//
// Text is what has arrived since the previous frame, so a client appends rather
// than replaces, and a backlog frame carrying the whole exchange so far appends
// into an empty run to the same effect. That is the point: one client-side code
// path whether the console was open when the generation started or not.
type llmFrame struct {
	Run     uint64 `json:"run"`
	VideoID string `json:"videoId"`
	Label   string `json:"label"`
	Model   string `json:"model"`
	Text    string `json:"text,omitempty"`
	Done    bool   `json:"done,omitempty"`
	// Error is the reason an exchange ended, present only alongside done. A
	// string rather than a problem document: this is a log, and what it is
	// logging has already been reported as a task failure through the API.
	Error string `json:"error,omitempty"`
	// Truncated says the text is a tail — output was dropped from the front to
	// bound what the server retains.
	Truncated bool      `json:"truncated,omitempty"`
	StartedAt time.Time `json:"startedAt"`
	// Millis is how long the exchange ran, present only once done.
	Millis int64 `json:"ms,omitempty"`
}

func toLLMFrame(f provider.LLMFrame) llmFrame {
	out := llmFrame{
		Run:       f.Run,
		VideoID:   f.Video.String(),
		Label:     f.Label,
		Model:     f.Model,
		Text:      f.Text,
		Done:      f.Done,
		Truncated: f.Truncated,
		StartedAt: f.StartedAt,
		Millis:    f.Duration.Milliseconds(),
	}
	if f.Err != nil {
		out.Error = f.Err.Error()
	}
	return out
}

// llmHandler streams what the language models are producing, right now.
//
// A second SSE endpoint rather than a kind on /events, and the split is the
// design. The event stream carries *state*: deltas are coalesced per video and
// merged last-wins, which is exactly right for a task that moved and exactly
// wrong for text, where merging two frames loses the first one's words. This
// carries an append-only log, so it gets its own connection with its own
// lifetime — opened when the console is shown and closed when it is hidden,
// costing nothing at all the rest of the time.
//
// There is no Last-Event-ID resume here for the same reason. The server retains
// the recent exchanges, so a reconnecting client is served by the ordinary
// backlog every client gets; an id to resume from would be a second answer to a
// question already answered.
func llmHandler(source LLMStreamSource, log *slog.Logger) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if source == nil {
			http.Error(w, "no llm log is configured", http.StatusServiceUnavailable)
			return
		}
		flusher, ok := w.(http.Flusher)
		if !ok {
			http.Error(w, "streaming unsupported", http.StatusInternalServerError)
			return
		}
		h := w.Header()
		h.Set("Content-Type", "text/event-stream")
		h.Set("Cache-Control", "no-cache, no-transform")
		h.Set("Connection", "keep-alive")
		// Disables proxy buffering where anything sits in front of the server.
		h.Set("X-Accel-Buffering", "no")
		w.WriteHeader(http.StatusOK)

		backlog, frames, cancel := source.Subscribe()
		defer cancel()

		for _, f := range backlog {
			if err := writeLLMFrame(w, f); err != nil {
				return
			}
		}
		flusher.Flush()

		ticker := time.NewTicker(heartbeatInterval)
		defer ticker.Stop()

		ctx := r.Context()
		for {
			select {
			case <-ctx.Done():
				return
			case f, open := <-frames:
				if !open {
					return
				}
				if err := writeLLMFrame(w, f); err != nil {
					log.Debug("llm console went away", slog.String("error", err.Error()))
					return
				}
				// Drain before flushing, so a burst costs one flush, not N.
			drain:
				for range 64 {
					select {
					case more, stillOpen := <-frames:
						if !stillOpen {
							flusher.Flush()
							return
						}
						if err := writeLLMFrame(w, more); err != nil {
							return
						}
					default:
						break drain
					}
				}
				flusher.Flush()
			case <-ticker.C:
				writeRaw(w, ": heartbeat\n\n")
				flusher.Flush()
			}
		}
	}
}

// writeLLMFrame emits one frame. No id field: nothing resumes from one here.
func writeLLMFrame(w http.ResponseWriter, f provider.LLMFrame) error {
	payload, err := json.Marshal(toLLMFrame(f))
	if err != nil {
		return err
	}
	if _, err := w.Write([]byte("event: llm\ndata: ")); err != nil {
		return err
	}
	if _, err := w.Write(payload); err != nil {
		return err
	}
	_, err = w.Write([]byte("\n\n"))
	return err
}
